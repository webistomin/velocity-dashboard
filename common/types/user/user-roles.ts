export enum UserRoles {
  ADMIN = 'admin',
  OPERATOR = 'operator',
}
