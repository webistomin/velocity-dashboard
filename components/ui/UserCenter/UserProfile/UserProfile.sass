+b('user-profile')
  padding: rem(64px) rem(24px)

  +e('heading')
    display: flex
    flex-direction: column
    align-items: center
    justify-content: center
    padding-bottom: rem(40px)
    margin-bottom: rem(40px)
    border-bottom: rem(1px) solid rgba(var(--color-primary), 0.2)

  +e('avatar-cropper')
    .ankaCropper.light
      font: var(--text-button)
      box-shadow: none
      background-color: rgb(var(--color-white))

      .ankaCropper__droparea
        border-color: rgb(var(--color-primary))

      a, button
        outline: none

      .ankaCropper__saveButton,
      .ankaCropper__selectButton
        display: flex
        align-items: center
        justify-content: center
        background-color: rgb(var(--color-primary))

        svg
          margin-right: rem(6px)

      .ankaCropper__selectButton
        margin: rem(24px) auto rem(10px)
        font: inherit
        font-weight: 500

      .ankaCropper__droparea
        div:first-child
          font: var(--text-caption)
          +letter-spacing(94px)
          text-transform: uppercase
          color: rgb(var(--color-gray-3))

      .ankaCropper__navigation
        box-shadow: none
        background-color: rgb(var(--color-white))
        border-bottom: rem(1px) solid rgb(var(--color-lavender))
        border-radius: 0

      .ankaCropper__navButton, .ankaCropper__saveButton
        box-shadow: none

      .ankaCropper__navButton svg
        stroke: rgb(var(--color-primary))

  +e('avatar-block')
    position: relative
    overflow: hidden
    border-radius: 50%
    margin-bottom: rem(24px)

    +media($screen-lg)
      &:hover .user-profile__avatar-changer
        opacity: 1
        visibility: visible

  +e('avatar-changer')
    position: absolute
    z-index: 1
    left: 0
    bottom: 0
    width: 100%
    height: 100%
    color: rgb(var(--color-white))
    background-color: rgba(var(--color-primary), 0.5)
    font: var(--text-button)
    border-radius: inherit
    transition: opacity 0.2s ease-out, visibility 0.2s ease-out, background-color 0.2s ease-out
    opacity: 0
    visibility: hidden

    &:active
      background-color: rgba(var(--color-primary), 0.7)

  +e('post')
    font-weight: 400
    color: rgb(var(--color-gray-3))
    margin-bottom: rem(35px)

  +e('actions')
    width: 100%
    display: flex
    justify-content: space-between
    align-items: center

    .base-button
      width: 48%

  +e('item')
    display: flex
    flex-direction: column

    &:not(:last-of-type)
      margin-bottom: rem(22px)

  +e('key')
    margin-bottom: rem(10px)

body.using-keyboard .ankaCropper *:focus
  outline: rem(1px) solid rgb(var(--color-primary))
