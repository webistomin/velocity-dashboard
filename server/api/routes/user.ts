import { Router, Request, Response } from 'express';
const route = Router();

export default (app: Router) => {
  app.use('/users', route);

  route.get('/me', (req: Request, res: Response) => {
    return res.json({ user: req }).status(200);
  });
};
