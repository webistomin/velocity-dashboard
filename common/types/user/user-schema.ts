import { Document, Types } from 'mongoose';
import { SiteThemes } from 'common/types/theme/site-themes';
import { UserRoles } from 'common/types/user/user-roles';
import { SentMessageInfo } from 'nodemailer';

export interface IUserInterface extends Document {
  firstName: string;
  lastName: string;
  role: UserRoles;
  email: string;
  password: string;
  theme: SiteThemes;
  notifications: {
    isEmailNotificationsEnabled: boolean;
    isPushNotificationsEnabled: boolean;
    isMonthlyNotificationsEnabled: boolean;
    isQuarterNotificationsEnabled: boolean;
  };
  avatar: string;
  phone: string;
  socials: {
    twitter: string;
  };
  location: string;
  bio: string;
}

export interface IUserSchema extends IUserInterface {
  __v: number;
  _id: Types.ObjectId;
  comparePassword(userPassword: string): Promise<boolean>;
  sendForgotPasswordMail(): Promise<SentMessageInfo>;
  sendSignUpMail(): Promise<SentMessageInfo>;
}
