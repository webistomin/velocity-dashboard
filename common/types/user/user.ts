import { IUserSchema } from './user-schema';

export interface IOwnUserProfileResponseBody {
  success: boolean;
  message?: string;
  user?: IUserSchema;
}
