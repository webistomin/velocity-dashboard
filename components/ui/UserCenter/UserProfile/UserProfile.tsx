import { VueComponent } from 'types/vue-components';
import { Component, Prop } from 'nuxt-property-decorator';
import { VNode } from 'vue';
// @ts-ignore
import AvatarCropper from 'vue-anka-cropper';
import { State } from 'vuex-class';

import { BaseThumbnail } from 'components/base/BaseThumbnail/BaseThumbnail';
import { BaseTitle } from 'components/base/BaseTitle/BaseTitle';
import BaseButton from 'components/base/BaseButton';
import { IUserInterface, IUserInterfaceDB } from 'common/types/user/user-schema';
import { UserRoles } from 'common/types/user/user-roles';
import BaseModal from 'components/base/BaseModal';
import { serverUrls } from 'common/urls/serverUrls';
import { BaseSpinner } from 'components/base/BaseSpinner/BaseSpinner';
import { IUploadAvatarResponseBody } from 'common/types/profile/avatar-upload';

import 'node_modules/vue-anka-cropper/dist/VueAnkaCropper.css';
import './UserProfile.sass';

export interface IUserProfileProps {
  info: IUserProfileInfo;
}

export interface IUserProfileInfo {
  firstName: IUserInterface['firstName'];
  lastName: IUserInterface['lastName'];
  email: IUserInterface['email'];
  location: IUserInterface['location'];
  bio: IUserInterface['bio'];
  avatar: IUserInterface['avatar'];
  role: IUserInterface['role'];
}

@Component({
  name: 'UserProfile',
  components: { AvatarCropper },
})
export default class UserProfile extends VueComponent<IUserProfileProps> {
  public isAvatarCropperVisible: boolean = false;
  public isLoading: boolean = false;

  @Prop()
  private readonly info!: IUserProfileProps['info'];

  public async logoutHandler(): Promise<void> {
    try {
      await this.$auth.logout();
    } catch (e) {
      this.$notify({
        group: 'common',
        type: 'error',
        title: 'Error during logout',
        text: e.message,
        duration: 3000,
      });
    }
  }

  public get getFullRoleName() {
    const role = this.info.role;

    switch (role) {
      case UserRoles.OPERATOR:
        return 'Operator';
      case UserRoles.ADMIN:
        return 'Administrator';
      default:
        return 'Not specified';
    }
  }

  public showAvatarCropper(): void {
    this.isAvatarCropperVisible = true;
  }

  public closeAvatarCropper(): void {
    this.isAvatarCropperVisible = false;
  }

  @State((state) => state.auth.user)
  getAuthUser!: IUserInterfaceDB;

  public async onAvatarUpload(data: any): Promise<void> {
    const newAvatarBlob = data.croppedFile;

    const formData = new FormData();
    formData.append('file', newAvatarBlob);
    formData.append('_id', this.getAuthUser._id);

    try {
      this.isLoading = true;
      const response: IUploadAvatarResponseBody = await this.$axios.$post(serverUrls.profile.uploadAvatar, formData);

      if (response.success) {
        await this.$auth.fetchUser();
        this.closeAvatarCropper();
        this.$notify({
          group: 'common',
          type: 'success',
          title: 'Success',
          text: 'Avatar has been successfully changed',
          duration: 3000,
        });
      }
      this.isLoading = false;
    } catch (error) {
      this.isLoading = false;
      const data: IUploadAvatarResponseBody = error.response.data;
      this.$notify({
        group: 'common',
        type: 'error',
        title: 'Success',
        text: data.message,
        duration: 3000,
      });
    }
  }

  render(): VNode {
    const { avatar, firstName, lastName, bio, location, email } = this.info;
    const fullName = `${firstName} ${lastName}`;

    return (
      <div class='user-profile'>
        <div class='user-profile__heading'>
          <BaseModal
            class='user-profile__avatar-modal'
            isVisible={this.isAvatarCropperVisible}
            v-scroll-lock={this.isAvatarCropperVisible}
            onClose={() => this.closeAvatarCropper()}>
            {this.isLoading ? (
              <BaseSpinner size='m' />
            ) : (
              <AvatarCropper
                class='user-profile__avatar-cropper'
                onCropper-saved={(cropData: any) => this.onAvatarUpload(cropData)}
                options={{
                  aspectRatio: 1,
                  closeOnSave: true,
                  cropArea: 'circle',
                  croppedHeight: 140,
                  croppedWidth: 140,
                  cropperHeight: false,
                  dropareaMessage: 'Drop file here or use the button below.',
                  frameLineDash: [5, 3],
                  frameStrokeColor: 'rgba(255, 255, 255, 0.8)',
                  handleFillColor: 'rgba(255, 255, 255, 0.2)',
                  handleHoverFillColor: 'rgba(255, 255, 255, 0.4)',
                  handleHoverStrokeColor: 'rgba(255, 255, 255, 1)',
                  handleSize: 10,
                  handleStrokeColor: 'rgba(255, 255, 255, 0.8)',
                  layoutBreakpoint: 850,
                  maxCropperHeight: 768,
                  maxFileSize: 8000000,
                  overlayFill: 'rgba(0, 0, 0, 0.5)',
                  previewOnDrag: true,
                  previewQuality: 0.65,
                  resultQuality: 0.8,
                  resultMimeType: 'image/jpeg',
                  selectButtonLabel: 'Select Files',
                  showPreview: true,
                  skin: 'light',
                  uploadData: {},
                  uploadTo: false,
                }}
              />
            )}
          </BaseModal>
          <div class='user-profile__avatar-block'>
            <button type='button' class='user-profile__avatar-changer btn' onClick={this.showAvatarCropper}>
              Change avatar
            </button>
            <BaseThumbnail id='profile-avatar' class='user-profile__avatar' image={avatar} alt={fullName} size='xl' />
          </div>
          <BaseTitle class='user-profile__title' level={3}>
            {fullName}
          </BaseTitle>
          <strong class='user-profile__post'>{this.getFullRoleName}</strong>
          <div class='user-profile__actions'>
            <BaseButton theme='light' onClick={this.logoutHandler}>
              Logout
            </BaseButton>
          </div>
        </div>
        <ul class='user-profile__list list'>
          <li class='user-profile__item list-item' vShow={email}>
            <strong class='user-profile__key caption'>Email</strong>
            <span class='user-profile__value'>{email}</span>
          </li>
          <li class='user-profile__item list-item' vShow={location}>
            <strong class='user-profile__key caption'>Location</strong>
            <span class='user-profile__value'>{location}</span>
          </li>
          <li class='user-profile__item list-item' vShow={bio}>
            <strong class='user-profile__key caption'>Bio</strong>
            <span class='user-profile__value'>{bio}</span>
          </li>
        </ul>
      </div>
    );
  }
}
