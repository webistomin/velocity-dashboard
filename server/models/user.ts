import mongoose, { Schema, HookNextFunction } from 'mongoose';
import bcrypt from 'bcrypt-nodejs';
import { IUserSchema } from 'common/types/user/user-schema';
import { SiteThemes } from 'common/types/theme/site-themes';

const UserSchema: Schema = new Schema({
  firstName: {
    type: String,
    required: true,
  },
  lastName: {
    type: String,
    required: true,
  },
  role: {
    type: String,
    required: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
  },
  password: {
    type: String,
    required: true,
  },
  theme: {
    type: String,
    default: SiteThemes.SHELOB,
  },
  notifications: {
    isEmailNotificationsEnabled: {
      type: Boolean,
      default: true,
    },
    isPushNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isMonthlyNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isQuarterNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
  },
  phone: {
    type: String,
    default: '',
  },
  socials: {
    twitter: {
      type: String,
      default: '',
    },
  },
  location: {
    type: String,
    default: '',
  },
  bio: {
    type: String,
    default: '',
  },
});

UserSchema.pre<IUserSchema>('save', function(next: HookNextFunction) {
  const user = this;
  if (this.isModified('password') || this.isNew) {
    bcrypt.genSalt(10, function(err, salt) {
      if (err) {
        return next(err);
      }

      bcrypt.hash(user.password, salt, null, function(err, hash) {
        if (err) {
          return next(err);
        }

        user.password = hash;
        next();
      });
    });
  } else {
    return next();
  }
});

UserSchema.methods.comparePassword = function(password: IUserSchema['password']): Boolean {
  const user = this;
  return bcrypt.compareSync(password, user.password);
};

export default mongoose.model<IUserSchema>('User', UserSchema);
