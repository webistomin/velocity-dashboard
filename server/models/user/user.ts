import { Schema, model, HookNextFunction } from 'mongoose';
import bcrypt from 'bcrypt';
import { nanoid } from 'nanoid';
import { SentMessageInfo } from 'nodemailer';
import consola from 'consola';

import { IUserInterface, IUserSchema } from 'common/types/user/user-schema';
import { SiteThemes } from 'common/types/theme/site-themes';
import PasswordReset from 'server/models/auth/password-reset';
import { sendResetMail } from 'server/services/mailer/reset-mail';

const UserSchema: Schema = new Schema<IUserInterface>({
  firstName: {
    type: String,
    required: true,
  },
  lastName: {
    type: String,
    required: true,
  },
  role: {
    type: String,
    required: true,
  },
  email: {
    type: String,
    required: true,
    unique: true,
  },
  password: {
    type: String,
    required: true,
  },
  theme: {
    type: String,
    default: SiteThemes.SHELOB,
  },
  notifications: {
    isEmailNotificationsEnabled: {
      type: Boolean,
      default: true,
    },
    isPushNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isMonthlyNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
    isQuarterNotificationsEnabled: {
      type: Boolean,
      default: false,
    },
  },
  phone: {
    type: String,
    default: '',
  },
  socials: {
    twitter: {
      type: String,
      default: '',
    },
  },
  location: {
    type: String,
    default: '',
  },
  bio: {
    type: String,
    default: '',
  },
});

UserSchema.pre<IUserSchema>('save', function(next: HookNextFunction) {
  const user = this;
  if (this.isModified('password') || this.isNew) {
    bcrypt.genSalt(10, function(err, salt) {
      if (err) {
        return next(err);
      }

      bcrypt.hash(user.password, salt, function(err, hash) {
        if (err) {
          return next(err);
        }

        user.password = hash;
        next();
      });
    });
  } else {
    return next();
  }
});

UserSchema.methods.comparePassword = async function(password: IUserSchema['password']): Promise<Boolean> {
  const user = this;
  const match = await bcrypt.compare(password, user.password);
  return match;
};

UserSchema.methods.forgotPassword = async function(): Promise<SentMessageInfo> {
  const token = nanoid(72);

  return await new Promise((resolve, reject) => {
    sendResetMail(this.email, token)
      .then(async (info) => {
        await PasswordReset.create({
          email: this.email,
          token,
          createdAt: new Date(),
        });
        resolve(info);
        consola.success({
          badge: true,
          message: `Reset mail successfully sent: ${info}`,
        });
      })
      .catch((error) => {
        reject(error);
        consola.error({
          badge: true,
          message: `Error during sending reset email: ${error}`,
        });
      });
  });
};

export default model<IUserSchema>('User', UserSchema);
