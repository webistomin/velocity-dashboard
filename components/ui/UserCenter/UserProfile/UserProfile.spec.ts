import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import { UserRoles } from 'common/types/user/user-roles';
import VScrollLock from 'v-scroll-lock';
import UserProfile from './UserProfile';

describe('UserProfile UI', () => {
  const localVue = createLocalVue();
  localVue.use(Vuex);
  localVue.use(VScrollLock);
  let store: any;
  let state;

  beforeEach(() => {
    state = {
      user: {
        firstName: 'Alexey',
        lastName: 'Istomin',
        email: 'mail@mail.ru',
        location: '',
        bio: '',
        avatar: '',
        role: UserRoles.ADMIN,
      },
    };

    store = new Vuex.Store({
      modules: {
        auth: {
          namespaced: true,
          state,
        },
      },
    });
  });

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      value: jest.fn(() => {
        return {
          matches: true,
          addEventListener: jest.fn(),
          removeEventListener: jest.fn(),
        };
      }),
      writable: true,
    });
  });

  const factory = (propsData = {}, options = {}) => {
    return shallowMount(UserProfile, {
      localVue,
      store,
      propsData: {
        info: {
          firstName: 'Alexey',
          lastName: 'Istomin',
          email: 'mail@mail.ru',
          location: '',
          bio: '',
          avatar: '',
          role: UserRoles.ADMIN,
        },
        ...propsData,
      },
      ...options,
    });
  };

  it('Matches snapshot', () => {
    const wrapper = factory().html();
    expect(wrapper).toMatchSnapshot();
  });
});
