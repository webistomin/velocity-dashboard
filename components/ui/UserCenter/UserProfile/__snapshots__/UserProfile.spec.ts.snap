// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile UI Matches snapshot 1`] = `
"<div class=\\"user-profile\\">
  <div class=\\"user-profile__heading\\">
    <basemodal-stub transitionname=\\"modal-fade\\" class=\\"user-profile__avatar-modal\\">
      <ankacropper-stub options=\\"[object Object]\\" class=\\"user-profile__avatar-cropper\\"></ankacropper-stub>
    </basemodal-stub>
    <div class=\\"user-profile__avatar-block\\"><button type=\\"button\\" class=\\"user-profile__avatar-changer btn\\">Change avatar</button>
      <vuecomponent-stub image=\\"\\" alt=\\"Alexey Istomin\\" size=\\"xl\\" id=\\"profile-avatar\\" class=\\"user-profile__avatar\\"></vuecomponent-stub>
    </div>
    <vuecomponent-stub level=\\"3\\" class=\\"user-profile__title\\">Alexey Istomin</vuecomponent-stub><strong class=\\"user-profile__post\\">Administrator</strong>
    <div class=\\"user-profile__actions\\">
      <basebutton-stub type=\\"button\\" icon=\\"\\" theme=\\"light\\">Logout</basebutton-stub>
    </div>
  </div>
  <ul class=\\"user-profile__list list\\">
    <li class=\\"user-profile__item list-item\\"><strong class=\\"user-profile__key caption\\">Email</strong><span class=\\"user-profile__value\\">mail@mail.ru</span></li>
    <li class=\\"user-profile__item list-item\\" style=\\"display: none;\\"><strong class=\\"user-profile__key caption\\">Location</strong><span class=\\"user-profile__value\\"></span></li>
    <li class=\\"user-profile__item list-item\\" style=\\"display: none;\\"><strong class=\\"user-profile__key caption\\">Bio</strong><span class=\\"user-profile__value\\"></span></li>
  </ul>
</div>"
`;
